src/directLineStreaming.ts

```diff
@@ -0,0 +1,351 @@
+      this.subscriber.error(new Error('there should be exactly one activity'));
+      const r = new BFSE.StreamingResponse();
+      r.statusCode = 500;
+      return r;
+    }
+
+    const activity = activitySet.activities[0];
+
+    if (streams.length > 0) {
+      let attachments = [...activity.attachments];
+
+      let stream: BFSE.ContentStream;
+      while (stream = streams.shift()) {
+        const attachment = await stream.readAsString();
+        const dataUri = "data:text/plain;base64," + attachment;
+        attachments.push({ contentType: stream.contentType, contentUrl: dataUri });
+      }
+
+      activity.attachments = attachments;
+    }
+
+    if (this.shouldQueue()) {
+      this.activityQueue.push(activity);
+    } else {
+      this.subscriber.next(activity);
+    }
+
+    const r = new BFSE.StreamingResponse();
+    r.statusCode = 200;
+    return r;
+  }
+
+  public flush() {
+    this.connectionStatus$.subscribe(cs => { })
+    this.activityQueue.forEach((a) => this.subscriber.next(a));
+    this.activityQueue = [];
+  }
+}
+
+export class DirectLineStreaming implements IBotConnection {
+  public connectionStatus$ = new BehaviorSubject(ConnectionStatus.Uninitialized);
+  public activity$: Observable<Activity>;
+
+  private activitySubscriber: Subscriber<Activity>;
+  private theStreamHandler: StreamHandler;
+
+  private retryCount = MAX_RETRY_COUNT;
+
+  private domain = "https://directline.botframework.com/v3/directline";
+
+  private conversationId: string;
+  private token: string;
+  private streamConnection: BFSE.WebSocketClient;
+  private queueActivities: boolean;
+
+  private _botAgent = '';
+
+  constructor(options: DirectLineStreamingOptions) {
+    this.token = options.secret || options.token;
+
+    if (options.token) {
+      this.refreshToken();
+    }
+
+    if (options.domain) {
+      this.domain = options.domain;
+    }
+
+    if (options.conversationId) {
+      this.conversationId = options.conversationId;
+    }
+
+    this._botAgent = this.getBotAgent(options.botAgent);
+
+    this.queueActivities = true;
+    this.activity$ = Observable.create(async (subscriber: Subscriber<Activity>) => {
+      this.activitySubscriber = subscriber;
+      this.theStreamHandler = new StreamHandler(subscriber, this.connectionStatus$, () => this.queueActivities);
+      await this.connectAsync();
+    }).share();
+  }
+
+  public reconnect(conversation: Conversation) {
+    this.conversationId = conversation.conversationId;
+    this.token = conversation.token;
+    this.connectAsync().then(_ => _);
+  }
+
+  end() {
+    this.connectionStatus$.next(ConnectionStatus.Ended);
+    this.streamConnection.disconnect();
+  }
+
+  private commonHeaders() {
+    return {
+      "Authorization": `Bearer ${this.token}`,
+      "x-ms-bot-agent": this._botAgent
+    };
+  }
+
+  private getBotAgent(customAgent: string = ''): string {
+    let clientAgent = 'directlinejs'
```

Comment:
`clientAgent` should denote if we're on DLSE or DLJS, right? `directlinejs` is not specific enough, to me.